File: spellcheck/spellcheck/LetterTree.cs
using System.Collections;
using System.Collections.Generic;
using System.Text;

namespace spellcheck
{
    public struct TraversalData
    {
        public int Depth;
        public LetterNode CurrentNode;
        public Dictionary<char,LetterNode> Nodes;
    }

    public class LetterTree
    {
        private const string NO_SUGGESTION_TEXT = "NO SUGGESTION";

        public Dictionary<char, LetterNode> Tree { get; private set; }
        public Dictionary<string, HashSet<int>> BadWord { get; private set; }

        public LetterTree()
        {
            Tree = new Dictionary<char,LetterNode>();
            BadWord = new Dictionary<string, HashSet<int>>();
        }

        public TraversalData GetRoot()
        {
            TraversalData treeData = new TraversalData();
            treeData.Depth = -1;
            treeData.CurrentNode = null;
            treeData.Nodes = Tree;

            return treeData;
        }

        public void Add(string word)
        {
            var traversal = Traverse(word, GetRoot());
            AddAtLocation(word, traversal);
        }

        private void AddAtLocation(string word, TraversalData traversal)
        {
            LetterNode node;

            if (traversal.Depth < word.Length - 1)
            {
                node = new LetterNode();
                node.Parent = traversal.CurrentNode;
                node.Letter = word[++traversal.Depth];
                traversal.Nodes.Add(node.Letter,node);

                if (traversal.Depth + 1 < word.Length)
                {
                    traversal.CurrentNode = node;
                    traversal.Nodes = node.Nodes;
                    AddAtLocation(word, traversal);
                }
                else
                    node.End = true;
            }
            else
                traversal.CurrentNode.End = true;
        }

        // Travels as far as it can before it either comes to the wrong letter
        // or comes to the end of the tree
        private TraversalData Traverse(string word, TraversalData traversal)
        {
            TraversalData oldSpot;
            TraversalData newSpot = traversal;

            do
            {
                oldSpot = newSpot;
                newSpot = TraverseOneStep(word, oldSpot);

            } while (oldSpot.Depth != newSpot.Depth);

            return newSpot;
        }

        private TraversalData TraverseOneStep(string word, TraversalData traversal)
        {
            var currentNode = TraverseNode(word, traversal);

            if (currentNode != null)
            {
                traversal.CurrentNode = currentNode;
                traversal.Nodes = currentNode.Nodes;
                traversal.Depth++;
            }

            return traversal;
        }


        private LetterNode TraverseNode(string word, TraversalData traversal)
        {
            if (word.Length > 0 && traversal.Depth < word.Length - 1)
            {
                var potentialKey = word[traversal.Depth + 1];

                // Check TraversalData.Nodes instead of TraversalData.CurrentNode
                // because there is not a single tree structure containing all of
                // the nodes. Instead it is a hash of nodes based on the first letter
                if (traversal.Nodes.ContainsKey(potentialKey))
                    return traversal.Nodes[potentialKey];
            }
            
            return null;
        }

        public string Spellcheck(string word)
        {
            var root = GetRoot();
            var newWord = Spellcheck(word, root);
            BadWord.Clear();
            return newWord;
        }

        // This function is called recursively from within itself and from within
        // each letter changing case (including case change)
        // The order is as follows:
        //   Go to the next letter in the word.
        //   If you can traverse to it, recursively call spellcheck at the next node in the tree
        //   Then if that returns no suggestion, check repeating letters, vowels, and improper casing.
        //   If any function returns a word that fits the requirements, stop immediately and return it
        //   If all options are exhausted, return no suggestion.
        public string Spellcheck(string word, TraversalData traversal,
                                 bool changingVowel = false, bool changingCase = false)
        {
            // If a word has been checked from this depth, don't check it again.
            if (!(BadWord.ContainsKey(word) && BadWord[word].Contains(traversal.Depth)))
            {
                if (traversal.CurrentNode != null)
                {
                    var currentWord = traversal.CurrentNode.GetWord();

                    if (traversal.CurrentNode.End == true && currentWord == word)
                        return currentWord;
                }

                var nextLocation = TraverseOneStep(word, traversal);
                if (traversal.Depth != nextLocation.Depth)
                {
                    var returnedWord = Spellcheck(word, nextLocation);

                    if (returnedWord != NO_SUGGESTION_TEXT)
                        return returnedWord;
                }

                // If a vowel was changed to cause the repeating character,
                // don't remove the repeated character.
                if ((traversal.Depth + 1 < word.Length && !IsVowel(word[traversal.Depth + 1]))
                    || !changingVowel)
                {
                    var processedWord = ProcessRepeatedLetter(word, traversal);

                    if (processedWord != NO_SUGGESTION_TEXT)
                        return processedWord;
                }

                // If you are still checking vowels, so don't try checking them again.
                if (!changingVowel)
                {
                    var changedVowel = CheckForIncorrectVowel(word, traversal, changingCase);

                    if (changedVowel != NO_SUGGESTION_TEXT)
                        return changedVowel;
                }

                // If you already changed the case, don't try checking it again.
                if (!changingCase)
                {
                    var changedCasing = CheckForImproperCasing(word, traversal, changingVowel);

                    if (changedCasing != NO_SUGGESTION_TEXT)
                        return changedCasing;
                }

                if (!BadWord.ContainsKey(word))
                    BadWord.Add(word, new HashSet<int>());
                BadWord[word].Add(traversal.Depth);
            }
            return NO_SUGGESTION_TEXT;
        }

        private string CheckForImproperCasing(string word, TraversalData traversal,
                                               bool changingVowel)
        {
            if (0 <= traversal.Depth + 1 && traversal.Depth + 1 < word.Length)
            {
                StringBuilder newWord = new StringBuilder(word);

                if (char.IsLower(word[traversal.Depth + 1]))
                    newWord[traversal.Depth + 1] = char.ToUpper(newWord[traversal.Depth + 1]);
                else
                    newWord[traversal.Depth + 1] = char.ToLower(newWord[traversal.Depth + 1]);

                return Spellcheck(newWord.ToString(), traversal, changingVowel, true);
            }
            return NO_SUGGESTION_TEXT;
        }

        private string CheckForIncorrectVowel(string word, TraversalData traversal,
                                              bool changingCase)
        {
            if (0 <= traversal.Depth + 1
                && traversal.Depth + 1 < word.Length
                && IsVowel(word[traversal.Depth + 1]))
            {
                StringBuilder newWord = new StringBuilder(word);

                foreach (char vowel in GetVowel())
                {
                    if (vowel != char.ToLower(word[traversal.Depth + 1]))
                    {
                        if (char.IsUpper(word[traversal.Depth + 1]))
                            newWord[traversal.Depth + 1] = char.ToUpper(vowel);
                        else
                            newWord[traversal.Depth + 1] = vowel;

                        var spellcheckReturn = Spellcheck(newWord.ToString(), traversal, true, changingCase);

                        if (spellcheckReturn != NO_SUGGESTION_TEXT)
                            return spellcheckReturn;
                    }
                }
            }
            return NO_SUGGESTION_TEXT;
        }

        private string ProcessRepeatedLetter(string word, TraversalData traversal)
        {
            // Strip off the repeating letter and run it through spellcheck again
            StringBuilder newWordBuilder = new StringBuilder(word);

            while (traversal.Depth + 2 < newWordBuilder.Length
                && newWordBuilder[traversal.Depth + 1] == newWordBuilder[traversal.Depth + 2])
            {
                newWordBuilder = newWordBuilder.Remove(traversal.Depth + 1, 1);

                string newWord = newWordBuilder.ToString();

                if (traversal.CurrentNode != null)
                {
                    var newTraversal = traversal;
                    newTraversal = TraverseOneStep(newWord, newTraversal);
                    newTraversal = TraverseOneStep(newWord, newTraversal);

                    if (traversal.Depth + 2 == newTraversal.Depth && newTraversal.CurrentNode.End
                        && newTraversal.CurrentNode.GetWord() == newWord)
                        return newTraversal.CurrentNode.GetWord();
                }
                var spellcheckReturn = Spellcheck(newWord, traversal);

                if (spellcheckReturn != NO_SUGGESTION_TEXT)
                    return spellcheckReturn;
            }
            return NO_SUGGESTION_TEXT;
        }

        private bool IsVowel(char character)
        {
            foreach (char vowel in GetVowel())
            {
                if (char.ToLower(character) == vowel)
                    return true;
            }
            return false;
        }

        // I used an array in the unspeller because I needed vowels
        // in a different way there. I could have used an array here
        // but I thought it would be fun trying the yield keyword
        // since I haven't really found a good use for yield yet.
        private IEnumerable GetVowel()
        {
            yield return 'a';
            yield return 'e';
            yield return 'i';
            yield return 'o';
            yield return 'u';
            yield return 'y';
        }
    }
}

File: RubySolution/lib/letter_tree.rb
require "#{File.dirname(__FILE__)}/traversal_data"
require "#{File.dirname(__FILE__)}/letter_node"
require "#{File.dirname(__FILE__)}/string_extension"
require "set"

class LetterTree
  @@NO_SUGGESTION_TEXT = "NO SUGGESTION"
  
  def initialize()
    @tree = {}
    @bad_words = {}
  end
  
  def get_root()
    treeData = TraversalData.new
    treeData.depth = -1
    treeData.current_node = nil
    treeData.nodes = @tree
    return treeData
  end
  
  def add(word)
    traversal = self.traverse(word, self.get_root())
    self.add_at_location(word, traversal)
  end
  
  def add_at_location(word, traversal)    
    if traversal.depth < word.length - 1
      traversal.depth += 1
      node = LetterNode.new
      node.parent = traversal.current_node
      node.letter = word[traversal.depth]
      traversal.nodes[node.letter] = node
      
      if traversal.depth + 1 < word.length
        traversal.current_node = node
        traversal.nodes = node.nodes
        add_at_location(word, traversal)
      else
        node.end = true
      end
    else
      traversal.current_node.end = true
    end
  end
  
  def traverse(word, traversal)
    begin
      oldSpot = traversal.depth
      traversal = self.traverse_one_step(word, traversal)
    end while traversal.depth != oldSpot
    
    return traversal
  end
  
  def traverse_one_step(word, traversal_ref)
    traversal = traversal_ref.clone
    current_node = self.traverse_node(word, traversal)

    if current_node != nil
      traversal.current_node = current_node
      traversal.nodes = current_node.nodes
      traversal.depth += 1
    end
    
    return traversal
  end
  
  def traverse_node(word, traversal)
    if word.length > 0 && traversal.depth < word.length - 1
      potential_key = word[traversal.depth+1]
      
      # Check traversal.nodes instead of traversal.current_node.nodes
      # because there is not a single tree structure containing all of
      # the nodes. Instead letter_tree has a has containing all of the
      # different roots. So to handle that case, traversal.nodes will
      # mimic traversal.current_node.nodes after the root. But at the
      # root it will have the root from get_root()
      if traversal.nodes.has_key?(potential_key)
        return traversal.nodes[potential_key]
      end
    end
  end
  
  def spellcheck(word)
    root = self.get_root()
    new_word = self.recursive_spellcheck(word, root)
    @bad_words.clear()
    return new_word
  end
  
  # Depth first check down the tree. If it can't find a word, backtrack and
  # process repeated letters, different vowels, and then casing.
  def recursive_spellcheck(word, traversal, changing_vowel = false,
                           changing_case = false)
    current_depth = traversal.depth
    # If the word is not in the badword list, or if the depth is not in the set of depth
    # checked for the bad word, run the spellcheck.
    if !@bad_words.has_key?(word) || !@bad_words[word].include?(current_depth)
      if traversal.current_node != nil
        current_word = traversal.current_node.get_word()
        
        if traversal.current_node.end == true && current_word == word
          return current_word
        end
      end
      
      next_location = self.traverse_one_step(word, traversal)
      if current_depth != next_location.depth
        returned_word = self.recursive_spellcheck(word, next_location)
        
        if returned_word != @@NO_SUGGESTION_TEXT
          return returned_word
        end
      end
      
      # If the next character is not a vowel, or if it is a vowel but we didn't
      # just change it, run spellcheck
      if (current_depth + 1 < word.length && !word[current_depth+1].is_vowel?) \
         || !changing_vowel
        returned_word = self.process_repeated_letter(word, traversal)
        
        if returned_word != @@NO_SUGGESTION_TEXT
          return returned_word
        end
      end
      
      if !changing_vowel
        returned_word = self.check_for_incorrect_vowel(word, traversal, changing_case)
        
        if returned_word != @@NO_SUGGESTION_TEXT
          return returned_word
        end
      end
      
      if !changing_case
        returned_word = self.check_for_improper_casing(word, traversal, changing_vowel)
        
        if returned_word != @@NO_SUGGESTION_TEXT
          return returned_word
        end
      end
      
      if !@bad_words.has_key?(word)
        @bad_words[word] = Set.new
      end
      @bad_words[word].add(current_depth)
    end
    return @@NO_SUGGESTION_TEXT
  end
  
  def check_for_improper_casing(word, traversal, changing_vowel)
    depth_to_check = traversal.depth + 1
    
    if 0 <= depth_to_check && depth_to_check < word.length
      new_word = word.clone
      new_word[depth_to_check] = new_word[depth_to_check].swapcase
      
      return self.recursive_spellcheck(new_word, traversal, changing_vowel, true)
    end
    return @@NO_SUGGESTION_TEXT
  end
  
  def check_for_incorrect_vowel(word, traversal, changing_case)
    depth_to_check = traversal.depth + 1
    
    if 0 <= depth_to_check && depth_to_check < word.length \
       && word[depth_to_check].is_vowel?
      new_word = word.clone
      # check each vowel but don't run spellcheck again on the passed in vowel
      String.Vowels.each do |vowel|
        if vowel != word[depth_to_check].downcase
          new_word[depth_to_check] = vowel.match_case(new_word[depth_to_check])
          
          returned_word = self.recursive_spellcheck(new_word, traversal, true, changing_case)
          
          if returned_word != @@NO_SUGGESTION_TEXT
            return returned_word
          end
        end
      end
    end
    return @@NO_SUGGESTION_TEXT
  end
  
  def process_repeated_letter(word, traversal)
    new_word = word.clone
    while traversal.depth + 2 < new_word.length \
          && new_word[traversal.depth + 1] == new_word[traversal.depth + 2]
      new_word.slice!(traversal.depth + 1)
      
      if traversal.current_node != nil
        new_traversal = traversal.clone
        new_traversal = self.traverse_one_step(new_word, new_traversal)
        new_traversal = self.traverse_one_step(new_word, new_traversal)
        
        if traversal.depth + 2 == new_traversal.depth && new_traversal.current_node.end \
           && new_traversal.current_node.get_word() == new_word
          return new_traversal.current_node.get_word()
        end
      end
      returned_word = self.recursive_spellcheck(new_word, traversal)
          
      if returned_word != @@NO_SUGGESTION_TEXT
        return returned_word
      end
    end 
    return @@NO_SUGGESTION_TEXT
  end
end
